File: Object-C/BlackFireIM/Example/Login/Controller/BFSignInStepTwoController.m
//
//  BFSignInStepTwoController.m
//  BlackFireIM
//
//  Created by benny wang on 2021/4/12.
//

#import "BFSignInStepTwoController.h"
#import "MSIMSDK-UIKit.h"
#import "BFRegisterInfo.h"
#import "AppDelegate.h"
#import "BFTabBarController.h"
#import "BFProfileService.h"
#import <TZImagePickerController.h>



@interface BFSignInStepTwoController()<TZImagePickerControllerDelegate>

@property(nonatomic,strong) UIImageView *avatarIcon;

@property(nonatomic,strong) UIButton *nextBtn;

@end
@implementation BFSignInStepTwoController

- (void)viewDidLoad
{
    [super viewDidLoad];
    UILabel *titleL = [[UILabel alloc]initWithFrame:CGRectMake(35, StatusBar_Height+NavBar_Height + 35, Screen_Width-70, 30)];
    titleL.text = @"MY PROFILE IS";
    titleL.font = [UIFont boldSystemFontOfSize:21];
    titleL.textColor = [UIColor d_colorWithColorLight:[UIColor blackColor] dark:[UIColor whiteColor]];
    [self.view addSubview:titleL];
    
    self.avatarIcon = [[UIImageView alloc]initWithFrame:CGRectMake(Screen_Width*0.5-112, titleL.maxY+40, 224, 218)];
    self.avatarIcon.contentMode = UIViewContentModeScaleAspectFill;
    self.avatarIcon.layer.cornerRadius = 8;
    self.avatarIcon.layer.masksToBounds = YES;
    self.avatarIcon.userInteractionEnabled = YES;
    self.avatarIcon.image = self.info.avatarImage ? self.info.avatarImage : [UIImage imageNamed:@"register_add"];
    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc]initWithTarget:self action:@selector(avatarDidTap)];
    [self.avatarIcon addGestureRecognizer:tap];
    [self.view addSubview:self.avatarIcon];
    
    self.nextBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    [self.nextBtn setTitle:TUILocalizableString(Next-button) forState:UIControlStateNormal];
    [self.nextBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    self.nextBtn.titleLabel.font = [UIFont systemFontOfSize:16];
    self.nextBtn.backgroundColor = RGBA(45, 45, 45, 1);
    self.nextBtn.layer.cornerRadius = 2;
    self.nextBtn.layer.masksToBounds = YES;
    [self.nextBtn addTarget:self action:@selector(nextBtnDidClick) forControlEvents:UIControlEventTouchUpInside];
    self.nextBtn.frame = CGRectMake(35, self.avatarIcon.maxY+60, Screen_Width-70, 50);
    [self.view addSubview:self.nextBtn];
}

- (void)avatarDidTap
{
    TZImagePickerController *picker = [[TZImagePickerController alloc]initWithMaxImagesCount:1 delegate:self];
    picker.allowPickingVideo = NO;
    picker.allowPickingImage = YES;
    [self presentViewController:picker animated:YES completion:nil];
}

- (void)nextBtnDidClick
{
    [MSHelper showToast];
    //先给个默认的头像，注册成功后再上传头像
    self.info.avatarUrl = @"https://msim-test-1252460681.cos.na-siliconvalley.myqcloud.com/pers/612FA7A3-144E-4978-A75C-9D9277167292.jpeg";
    [self reqeustToSignUp];
}

- (void)reqeustToSignUp
{
    WS(weakSelf)
    [BFProfileService userSignUp:self.info succ:^{
        [BFProfileService requestIMToken:weakSelf.info.phone success:^(NSDictionary * _Nonnull dic) {
            
            NSString *userToken = dic[@"token"];
            NSString *im_url = dic[@"url"];
            weakSelf.info.userToken = userToken;
            weakSelf.info.imUrl = im_url;
            // 配置聊天室
            [MSChatRoomManager.sharedInstance loginChatRoom:kChatRoomID];
            [[MSIMManager sharedInstance] login:weakSelf.info.userToken imUrl:weakSelf.info.imUrl subAppID:1 succ:^{
                            
                STRONG_SELF(strongSelf)
                            AppDelegate *appDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
                            appDelegate.window.rootViewController = [[BFTabBarController alloc]init];
                            
                            [strongSelf uploadAvatar];
                        } failed:^(NSInteger code, NSString *desc) {
                            [MSHelper showToastFail:desc];
            }];
            
        } fail:^(NSError * _Nonnull error) {
            [MSHelper showToastFail:error.localizedDescription];
        }];
    } failed:^(NSError * _Nonnull error) {
        [MSHelper showToastFail:error.localizedDescription];
    }];
}

- (void)uploadAvatar
{
    if (self.info.avatarImage == nil) return;
    MSIMImageElem *elem = [[MSIMImageElem alloc]init];
    elem.image = self.info.avatarImage;
    [[MSIMManager sharedInstance].uploadMediator ms_uploadWithObject:elem.image fileType:MSUploadFileTypeAvatar progress:^(CGFloat progress) {

    } succ:^(NSString * _Nonnull url) {

        MSProfileInfo *info = [[MSProfileProvider provider]providerProfileFromLocal:[MSIMTools sharedInstance].user_id];
        info.avatar = url;
        NSMutableDictionary *dic = [info.custom el_convertToDictionary].mutableCopy;
        dic[@"pic"] = url;
        info.custom = [dic el_convertJsonString];
        [BFProfileService requestToEditProfile:info success:^(NSDictionary * _Nonnull dic) {
            
            [[MSProfileProvider provider]updateProfiles:@[info]];
            
        } fail:^(NSError * _Nonnull error) {
            
        }];

    } fail:^(NSInteger code, NSString * _Nonnull desc) {

    }];
}

#pragma mark - TZImagePickerControllerDelegate

- (void)imagePickerController:(TZImagePickerController *)picker didFinishPickingPhotos:(NSArray<UIImage *> *)photos sourceAssets:(NSArray *)assets isSelectOriginalPhoto:(BOOL)isSelectOriginalPhoto infos:(NSArray<NSDictionary *> *)infos
{
    self.avatarIcon.image = photos.firstObject;
    self.info.avatarImage = photos.firstObject;
}

@end

File: Swift/MSIM-Swift/Login/BFSignInStepTwoController.swift
//
//  BFSignInStepTwoController.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/7/28.
//

import UIKit
import Photos
import ZLPhotoBrowser
import MSIMSDK


class BFSignInStepTwoController: BFBaseViewController {
    
    var info: BFRegisterInfo!
    
    private var avatarIcon: UIImageView!
    
    private var nextBtn: UIButton!
    
    override func viewDidLoad() {
        super.viewDidLoad()

        let titleL = UILabel(frame: CGRect(x: 35, y: UIScreen.status_navi_height + 35, width: UIScreen.width - 70, height: 30))
        titleL.text = "MY PROFILE IS"
        titleL.font = .boldSystemFont(ofSize: 21)
        titleL.textColor = UIColor.d_color(light: .black, dark: .white)
        view.addSubview(titleL)
        
        avatarIcon = UIImageView(frame: CGRect(x: UIScreen.width * 0.5 - 112, y: titleL.bottom + 40, width: 224, height: 218))
        avatarIcon.contentMode = .scaleAspectFill
        avatarIcon.layer.cornerRadius = 8
        avatarIcon.layer.masksToBounds = true
        avatarIcon.isUserInteractionEnabled = true
        avatarIcon.image = info.avatarImage ?? UIImage(named: "register_add")
        let tap = UITapGestureRecognizer(target: self, action: #selector(avatarDidTap))
        avatarIcon.addGestureRecognizer(tap)
        view.addSubview(avatarIcon)
        
        nextBtn = UIButton(type: .custom)
        nextBtn.setTitle(Bundle.bf_localizedString(key: "Next-button"), for: .normal)
        nextBtn.setTitleColor(.white, for: .normal)
        nextBtn.titleLabel?.font = .systemFont(ofSize: 16)
        nextBtn.backgroundColor = UIColor(r: 45, g: 45, b: 45)
        nextBtn.layer.cornerRadius = 2
        nextBtn.layer.masksToBounds = true
        nextBtn.addTarget(self, action: #selector(nextBtnDidClick), for: .touchUpInside)
        nextBtn.frame = CGRect(x: 35, y: avatarIcon.bottom + 60, width: UIScreen.width - 70, height: 50)
        view.addSubview(nextBtn)
        
    }
    

    @objc func avatarDidTap() {
        
        let config = ZLPhotoConfiguration.default()
        config.allowSelectVideo = false
        config.allowSelectImage = true
        config.maxSelectCount = 1
        config.allowEditImage = true
        config.allowEditVideo = false
        config.imageStickerContainerView = ImageStickerContainerView()
        
        config.canSelectAsset = { (asset) -> Bool in
            return true
        }
        
        config.noAuthorityCallback = { (type) in
            switch type {
            case .library:
                debugPrint("No library authority")
            case .camera:
                debugPrint("No camera authority")
            case .microphone:
                debugPrint("No microphone authority")
            }
        }
        
        let picker = ZLPhotoPreviewSheet(selectedAssets: nil)
        picker.selectImageBlock = {[weak self] (images,assets,isOriginal) in
            self?.didPickerAsset(image: images.first!,asset: assets.first!)
        }
        picker.cancelBlock = {
            
        }
        picker.selectImageRequestErrorBlock = { (errorAssets, errorIndexs) in
            debugPrint("fetch error assets: \(errorAssets), error indexs: \(errorIndexs)")
        }
        picker.showPhotoLibrary(sender: self)
    }
    
    @objc func nextBtnDidClick() {
        
        if info.avatarImage == nil {
            MSHelper.showToastWithText(text: "请上传头像")
            return
        }
        MSHelper.showToast()
        let elem = MSIMImageElem()
        elem.image = info.avatarImage!
        MSIMManager.sharedInstance().uploadMediator?.ms_upload?(with: elem.image!, fileType: .avatar, progress: { progress in
            
        }, succ: { url in
            
            self.info.avatarUrl = url
            self.requestToSignUp()
            
        }, fail: { code, desc in
            
            MSHelper.showToastFailWithText(text: desc)
        })
    }
    
    private func requestToSignUp() {
        ProfileService.userRegistAPI(phone: info.phone!, nickName: info.nickName!, avatar: info.avatarUrl!) { _ in
            
            ProfileService.iMTokenAPI(uid: self.info.phone!) { result in
                
                let dic = result as! [String: Any]
                let userToken = dic["token"] as! String
                let im_Url = dic["url"] as! String
                self.info.userToken = userToken
                self.info.imUrl = im_Url
                MSIMManager.sharedInstance().login(userToken, imUrl: im_Url, subAppID: 1) {
                    
                    let appDelegate = UIApplication.shared.delegate as! AppDelegate
                    appDelegate.window?.rootViewController = BFTabBarController()
                    
                } failed: { code, desc in
                    MSHelper.showToastFailWithText(text: desc ?? "")
                }
            } fail: { error in
                MSHelper.showToastFailWithText(text: error?.localizedDescription ?? "")
            }
        } fail: { error in
            MSHelper.showToastFailWithText(text: error?.localizedDescription ?? "")
        }
    }
}

extension BFSignInStepTwoController {
    
    private func didPickerAsset(image: UIImage,asset: PHAsset) {
        
        self.avatarIcon.image = image
        self.info.avatarImage = image
    }
}
